[DOC-ADD] Use Case : XML_ID renaming (#3844)
OCA/OpenUpgrade: docsource/use_cases/xml_id_renaming.rst

> XML_ID Renaming

Reviewer: ```suggestion
XML-ID Renaming
```

+++++++++++++++

> Exemple

Reviewer: ```suggestion
Example
```

.. code-block:: text

    ---Fields in module 'crm'---
    DEL ir.actions.act_window.view: crm.action_crm_tag_form_view_oppor11
    NEW ir.actions.act_window.view: crm.crm_lead_opportunities_view_form
    # DONE: pre-migration: Renamed XML-IDs

See `Full V12 Work Analysis File <https://github.com/OCA/OpenUpgrade/blob/568c9c5c35f443df9c2d0a83b869f9fb5a6de737/addons/crm/migrations/12.0.1.0/openupgrade_analysis_work.txt#L38>`_.

Write migration Script
""""""""""""""""""""""

in the ``pre-migration.py`` script add:

.. code-block:: python

    from openupgradelib import openupgrade

    _xml_ids_renames = [
        ('crm.action_crm_tag_form_view_oppor11',
         'crm.crm_lead_opportunities_view_form'),
    ]

    @openupgrade.migrate()
    def migrate(env, version):
        openupgrade.rename_xmlids(env.cr, _xml_ids_renames)

See `Full pre migration Script <https://github.com/OCA/OpenUpgrade/blob/568c9c5c35f443df9c2d0a83b869f9fb5a6de737/addons/crm/migrations/12.0.1.0/pre-migration.py#L24>`_.

Notes
-----

* This operation is placed in the ``pre-migration.py`` step, to be executed before the
  load of the data, in the regular update process.

> * In most cases, the xml_id renaming is required because from a version to the next one,

Reviewer: ```suggestion
* In most cases, the XML-ID renaming is required because from a version to the next one,
```

  Odoo moves data from a module to another.

  For exemple, from 13.0 to 14.0 some data have been moved from ``sale_timesheet``
  to ``sale_project``.

  See : `Sale Project migration script <https://github.com/OCA/OpenUpgrade/blob/6fded25b7914ee3c1e5f3066252807e7f9b92cbe/openupgrade_scripts/scripts/sale_project/14.0.1.0/pre-migration.py#L1>`_.

* If the data is flagged as ``noupdate="1"`` the table after the upgrade
  (without migration script) will contains a duplicates.

  This duplicates can generate a crash of your upgrade, if some constraints are present.

>   For exemple, the ``ir.config.parameter`` named ``mail.icp_mail_bounce_alias`` (V14) has been

Reviewer: ```suggestion
  For example, the ``ir.config.parameter`` named ``mail.icp_mail_bounce_alias`` (V14) has been
```

  renamed into ``base.icp_mail_bounce_alias`` (V15).
  As the ``ir.config.parameter`` is created with a ``noupdate="1"`` flag AND there is
  a unicity constraint in the ``key`` field, the creation of the duplicates will fail, during the upgrade
  process.

* In the case of merging a module, the renaming is done in two steps:

>   For exemple, V12 module ``mrp_byproduct`` has been merged into ``mrp`` module in V13.

Reviewer: ```suggestion
  For example, the v12 module ``mrp_byproduct`` has been merged into ``mrp`` module in v13.
```
